step-capstone/src/components/Map.js

```diff
@@ -6,37 +6,215 @@ import React, { createRef } from 'react'
+    if (prevProps !== this.props) {
+      this.clearMap();
+      this.drawMap();
+    }
+  }
+
   createMap() {
-    new window.google.maps.Map(this.googleMapRef.current, {
+    return new window.google.maps.Map(this.googleMapRef.current, {
       zoom: this.props.zoom,
       center: this.props.center,
     })
   }
 
+  addMarker(coordinates, type, label) {
+    // TODO: firebase will provide coordinates as a GeoPoint --> convert to {lat, lng}
+    // { lat: coordinates.lat(), lng: coordinates.lng() }
+
+    var iconUrl;
+    if (type === "flight") {
+      iconUrl = "http://maps.google.com/mapfiles/ms/icons/blue-dot.png"
+    } else if (type === "event") {
+      iconUrl = "http://maps.google.com/mapfiles/ms/icons/green-dot.png"
+    } else {
+      iconUrl = "http://maps.google.com/mapfiles/ms/icons/red-dot.png"
+    }
+
+    var newMarker = new window.google.maps.Marker({
+      position: coordinates,
+      map: this.googleMap,
+      animation: window.google.maps.Animation.DROP,
+      icon: { url: iconUrl }
+    })
+
+    // zoom to marker when clicked
+    newMarker.addListener('click', () => {
+      this.googleMap.setZoom(15);
+      this.googleMap.setCenter(newMarker.getPosition());
+
+      // TODO: this.props.onMarkerClicked(id) --> notifies parent that marker clicked
+    });
+
+    return newMarker;
+  }
+
+  // constructs path object and places on map. Returns path object.
+  addPath(map, path, color) {
+    // TODO : change to show directions between paths
+    // dotted line
+    var lineSymbol = {
+      path: 'M 0,-1 0,1',
+      strokeOpacity: 1,
+      scale: 4
+    };
+    
+    let geoPath = new window.google.maps.Polyline({
+      path: path,
+      strokeColor: color,
+      strokeOpacity: 0,
+      strokeWeight: 5,
+      icons: [{
+          icon: lineSymbol,
+          offset: '0',
+          repeat: '20px'
+        }]
+    });
+
+    geoPath.setMap(map);
+
+    return geoPath;
+  }
+
+  /* 
+   * Given a list of travel object data, draws corresponding markers on map and 
+   * returns hash map containing marker objects 
+   */
+  drawMarkers(list, bounds) {
+    // construct hashmap with key: travelObject id, value: marker object
+    return list.reduce((hashMap, item) => {
+      if (item.type !== "flight") {
+        hashMap.set(item.id, {marker: this.addMarker(item.coordinates, item.type, item.id), type: item.type});
+        bounds.extend(item.coordinates);
+      } else {
+        hashMap.set(item.id, {
+          marker: {
+            departure: this.addMarker(item.departureCoordinates, item.type, item.id),
+            arrival: this.addMarker(item.arrivalCoordinates, item.type, item.id)
+          }, type: item.type});
+
+        bounds.extend(item.departureCoordinates);
+        bounds.extend(item.arrivalCoordinates);
+      }
+      return hashMap;
+    }, new Map())
+  }
+
+  /*
+   *  Draws path between all finalized travel objects, returns list of path objects
+   */
+  drawPaths() {
```

Comment:
This function ended up being quite long. I would consider separating the color helper into a little separate function to help with the length.